doc/design/program.md

```diff
@@ -2,7 +2,7 @@
 
 ## Compile and Execution
 
-A PaddlePaddle program consists of two parts -- the first generates a `ProgramDesc` protobuf message that describes the program, and the second runs this message using a C++ class `Executor`.
+A PaddlePaddle program consists of three parts -- the first generates a `ProgramDesc` protobuf message that describes the program, the second optimizes this message using a C++ class `Optimizer` and generates an `ExecutionPlan` protobuf messages, and the third run the message using a C++ class `Executor`.
```

Comment:
The same as @dzhwinter , just my personal view, this **Optimizer** does not do the **optimize**,  like the four steps which run a C program, COMPILER -> ASSEMBLER -> LINKER -> LOADER. How about convert `optimizer` -> `assembler`?

Reply:
Generating `ExecutionPlan` is exactly like `gcc`'s `-O` option.

We probably do not need a single C++ class to optimize the graph since we can just create a member function `OptimizeProgram` in the `Executor` class. `ExecutionPlan` object should also be the member of `Executor`, so that we call `executor.run` is executing optimized graph.

Reply:
And, `ExecutionPlan` is no need to be a protobuf.

Reply:
Please refer to #6141 , I think that `ProgramDesc` is not enough to run a network. We also need to provide Device Type and Data Type for each operator. Exposing these interface to users is necessary, even though paddle framework could provide a solution.

Reply:
@dzhwinter @Yancey1989 
> We'd better switch Optimizer to another term. Our python already have the Optimizer.

Agree we need a better naming, thank @Yancey1989 ! Assembler is a good name candidate!

@typhoonzero I think whoever generates the `ExecutionPlan` from `ProgramDesc` should have the global view: the global program desc, and the number of devices. And different `ExecutionPlans` are sent to different nodes. On the other hand, `Executor` runs locally, it does not know the devices on other nodes.

Reply:
@zealoct thank you! I have changed the name to `Planner`, do you think it conveys the means correctly?